cosmos/precompile/bank/bank.go

```diff
@@ -21,26 +21,395 @@
+	return []any{supply.BigInt()}, nil
+}
+
+// GetTotalSupply implements `getTotalSupply()` method.
+func (c *Contract) GetTotalSupply(
+	ctx context.Context,
+	_ ethprecompile.EVM,
+	_ common.Address,
+	_ *big.Int,
+	readonly bool,
+	args ...any,
+) ([]any, error) {
+	// todo: add pagination here
+	res, err := c.querier.TotalSupply(ctx, &banktypes.QueryTotalSupplyRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	return []any{sdkCoinsToEvmCoins(res.Supply)}, nil
+}
+
+// GetDenomMetadata implements `getDenomMetadata(string)` method.
+func (c *Contract) GetDenomMetadata(
+	ctx context.Context,
+	_ ethprecompile.EVM,
+	_ common.Address,
+	_ *big.Int,
+	readonly bool,
+	args ...any,
+) ([]any, error) {
+	denom, ok := utils.GetAs[string](args[0])
+	if !ok {
+		return nil, precompile.ErrInvalidString
+	}
+
+	res, err := c.querier.DenomMetadata(ctx, &banktypes.QueryDenomMetadataRequest{
+		Denom: denom,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	r := res.Metadata
+
+	denomUnits := make([]generated.IBankModuleDenomUnit, len(r.DenomUnits))
+	for i, d := range r.DenomUnits {
+		denomUnits[i] = generated.IBankModuleDenomUnit{
+			Denom:    d.Denom,
+			Aliases:  d.Aliases,
+			Exponent: d.Exponent,
+		}
+	}
+
+	result := generated.IBankModuleDenomMetadata{
+		Description: r.Description,
+		DenomUnits:  denomUnits,
+		Base:        r.Base,
+		Display:     r.Display,
+		Name:        r.Name,
+		Symbol:      r.Symbol,
+	}
+	return []any{result}, nil
+}
+
+// GetSendEnabled implements `getSendEnabled(string[])` method.
+func (c *Contract) GetSendEnabled(
+	ctx context.Context,
+	_ ethprecompile.EVM,
+	_ common.Address,
+	_ *big.Int,
+	readonly bool,
+	args ...any,
+) ([]any, error) {
+	denom, ok := utils.GetAs[string](args[0])
+	if !ok {
+		return nil, precompile.ErrInvalidString
+	}
+
+	res, err := c.querier.SendEnabled(ctx, &banktypes.QuerySendEnabledRequest{
+		Denoms: []string{denom},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(res.SendEnabled) == 0 {
+		return nil, precompile.ErrInvalidString
+	}
+
+	return []any{res.SendEnabled[0].Enabled}, nil
+}
+
+// msg_server functions
+// Send implements `send(address,address,(uint256,string))` method.
+func (c *Contract) Send(
+	ctx context.Context,
+	_ ethprecompile.EVM,
+	_ common.Address,
+	_ *big.Int,
+	readonly bool,
+	args ...any,
+) ([]any, error) {
+	fromAddr, ok := utils.GetAs[common.Address](args[0])
+	if !ok {
+		return nil, precompile.ErrInvalidHexAddress
+	}
+	toAddr, ok := utils.GetAs[common.Address](args[1])
+	if !ok {
+		return nil, precompile.ErrInvalidHexAddress
+	}
+	coins, err := extractCoinsFromInput(args[2])
+
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = c.msgServer.Send(ctx, &banktypes.MsgSend{
+		FromAddress: cosmlib.AddressToAccAddress(fromAddr).String(),
+		ToAddress:   cosmlib.AddressToAccAddress(toAddr).String(),
+		Amount:      coins,
+	})
+	return []any{err == nil}, err
+}
+
+// note: not doing it now, it causes too much trouble
```

Comment:
github issue, remove the commented out code from the repo, and paste the commented code out in the github issue, reference filename and stuff